"Dynamic atom creation" added (Anti-patterns documentation) (#12804)
elixir-lang/elixir: lib/elixir/pages/anti-patterns/code-anti-patterns.md

## Dynamic atom creation

#### Problem

> An `atom` is a primary data type of Elixir whose value is its own name. They are often useful to identify resources or express an operation's state. The creation of an `atom` do not characterize an anti-pattern by itself; however, `atoms` are not collected by Elixir's Garbage Collector, so values of this type live in memory while software is executing, during its entire lifetime. Also, BEAM limits the number of `atoms` that can exist in an application (`1_048_576`) and each `atom` has a maximum size limited to 255 Unicode code points. For these reasons, the dynamic atom creation is considered an anti-pattern, since in this way the developer has no control over how many `atoms` will be created during the software execution. This unpredictable scenario can expose the software to unexpected behavior caused by excessive memory usage, or even by reaching the maximum number of `atoms` possible.

Reviewer: We shouldn't use backquotes for things that are not code, because it's not semantically right (and hinders accessibility as well 😢). Can you replace all the uses of `` `atom` `` and friends ("string" for example) with italic (`*atom*`) instead?

Reviewer 2: Agreed. If we are going to use backquote, it should be `Atom`, but it should be used once or twice, and not on every reference.

Author: OK! I agree. I just made a commit to fix it. I'm sorry for that mistake... :)

> An `atom` is a primary data type of Elixir whose value is its own name. They are often useful to identify resources or express an operation's state. The creation of an `atom` do not characterize an anti-pattern by itself; however, `atoms` are not collected by Elixir's Garbage Collector, so values of this type live in memory while software is executing, during its entire lifetime. Also, BEAM limits the number of `atoms` that can exist in an application (`1_048_576`) and each `atom` has a maximum size limited to 255 Unicode code points. For these reasons, the dynamic atom creation is considered an anti-pattern, since in this way the developer has no control over how many `atoms` will be created during the software execution. This unpredictable scenario can expose the software to unexpected behavior caused by excessive memory usage, or even by reaching the maximum number of `atoms` possible.

Reviewer: Some notes:

1. We refer to "primary data types" as "basic data types" in https://hexdocs.pm/elixir/typespecs.html, so might as well do the same here.

```suggestion
An `atom` is a basic data type of Elixir whose value is its own name. Atoms are often useful to identify resources or express the state or result of an operation. Creating atoms dynamically is not an anti-pattern by itself; however, `atoms` are not collected by the BEAM's garbage collector, so values of this type live in memory while software is executing, during its entire lifetime. Also, BEAM limits the number of `atoms` that can exist in an application (`1_048_576` by default) and each `atom` has a maximum size limited to 255 Unicode code points. For these reasons, creating atoms dynamically is considered an anti-pattern, since in this way the developer has no control over how many `atoms` will be created during the software execution. This unpredictable scenario can expose the software to unexpected behavior caused by excessive memory usage, or even by reaching the maximum number of `atoms` possible.
```

#### Example

Imagine that you are implementing a code that performs the conversion of `string` values into `atoms` to identify resources. These `strings` can come from user input or even have been received as response from requests to an API. As this is a dynamic and unpredictable scenario, it is possible for identical `strings` to be converted into new `atoms` that are repeated unnecessarily. This kind of conversion, in addition to wasting memory, can be problematic for the software if it happens too often.

```elixir
defmodule Identifier do

>   def generate(id) when is_bitstring(id) do

Reviewer: ```suggestion
  def generate(id) when is_binary(id) do
```

    String.to_atom(id)  #<= dynamic atom creation!!
  end
end
```

```elixir
iex> string_from_user_input = "my_id"
"my_id"

> iex> string_from_API_response = "my_id"

Reviewer: ```suggestion
iex> string_from_api_response = "my_id"
```

"my_id"
iex> Identifier.generate(string_from_user_input)
:my_id
iex> Identifier.generate(string_from_API_response)
:my_id   #<= atom repeated was created!
```

> When we use the `String.to_atom/1` function to dynamically create an `atom`, it is created regardless of whether there is already another one with the same value in memory, so when this happens automatically, we will not have control over meeting the limits established by BEAM.

Reviewer: I'm going to expose my ignorance here... I don't think this is true? The docs for `String.to_atom/1` and `:erlang.binary_to_atom/1` don't mention anything like this, and it goes against the definition of atoms themselves, so I'd be surprised if this was the case 🤔

Author: > I confirm it is not the case. Existing atoms are not created again.

Curious. In my understanding, this is an important point in the "root" of this anti-pattern definition. @josevalim, do you have any suggestions on how to modify this in our text?

Reviewer 2: The issue about String.to_atom is for atoms that do not yet exist. If you are covering an external parameter, such as requests from an API to atom, then someone can use to create enough atoms to cause the system to shut down. Which is why @whatyouhide mentioned APIs being a key issue.

#### Refactoring

To remove this anti-pattern, we must ensure that all the identifier *atoms* are created statically, only once, at the beginning of an application's execution:

```elixir
# statically created atoms...
_ = :my_id
_ = :my_id2
_ = :my_id3
_ = :my_id4

> ```

Reviewer 3: I've never thought about doing this, what would be the right place to do this? `application.ex`?

Reviewer: This is not a common pattern in my experience though.

Reviewer 2: This advice is not correct as well. The atoms must be defined on the module that attempts to call `String.to_existing_atom`, because code is loaded at module-level, not at an application level. The notes here are up to date: https://hexdocs.pm/elixir/String.html#to_existing_atom/1 :)

Next, you should replace the use of the `String.to_atom/1` function with the `String.to_existing_atom/1` function. This will allow string-to-atom conversions to just map the strings to atoms already in memory (statically created at the beginning of the execution), thus preventing repeated *atoms* from being created dynamically. This second part of the refactoring is presented below.

```elixir
defmodule Identifier do
  def generate(id) when is_binary(id) do

>     String.to_existing_atom(id)  #<= just maps a string to an existing atom!

Reviewer 3: Another strategy I've often used instead of `String.to_existing_atom/1` was to pattern-match only on an expected subset, which works as validation:

```elixir
def generate("foo"), do: :foo
def generate("bar"), do: :bar
def generate("baz"), do: :baz
```

(or wrap `def` in the `for` in bigger cases, or use `Ecto.Enum` when applies).
I wonder if we should mention this strategy too, or if it would be too much noise.

Reviewer: This is 100% what I tell folks to do most of the time in place of using `String.to_existing_atom/1`. I think it'd be great to suggest it over `to_existing_atom/1`, and then only mention `to_existing_atom/1` as an alternative.

  end
  
  def generate_atom(response) when is_map(response)
    String.to_atom(response.type)  #<= dynamic atom creation!!
  end
end
```

```elixir
iex> API.request("bar") |> API.generate_atom()
:error
iex> API.request("qux") |> API.generate_atom()
:ok
```

When we use the `String.to_atom/1` function to dynamically create an *atom* in an API, it essentially gains potential access to create arbitrary *atoms* in our system, causing us to lose control over adhering to the limits established by the BEAM. This issue could be exploited by someone to create enough *atoms* to shut down a system.

#### Refactoring

To eliminate this anti-pattern, before any string-to-atom conversion takes place, we must ensure that all potential *atoms* to be created from *string* conversions are statically defined in the module where these conversions will occur. Next, you should replace the use of the `String.to_atom/1` function with the `String.to_existing_atom/1` function. This will allow us to have greater control over which *atoms* can be created by the API, as only those *atoms* loaded at the module level will be convertible from *strings*.

```elixir
defmodule API do
  # statically created atoms

>   _ = :error

Reviewer 4: Using this syntax might be confusing, what if these were module attributes instead? Or a single list of one module attribute?

Author: > Using this syntax might be confusing, what if these were module attributes instead? Or a single list of one module attribute?

Something like this @atoms [:error, :ok]?

Reviewer 2: None of those work. The atoms must be defined in a function, anything in the module body is compiled out as well. Anyway, I think defining atoms like this can be an anti-pattern too. The best is to call `to_existing_atom` close to where the atom is used. I will review it fully soon. :)

Author: > None of those work. The atoms must be defined in a function, anything in the module body is compiled out as well. Anyway, I think defining atoms like this can be an anti-pattern too. The best is to call `to_existing_atom` close to where the atom is used. I will review it fully soon. :)

@josevalim Considering your comment, I made another adjustment to the documentation by adding the private function `static_atom_creation()`. It is called just before the calls to `String.to_existing_atom()` in functions of the module used in the example.

## Dynamic atom creation

#### Problem

> An `Atom` is a basic data type of Elixir whose value is its own name. *Atoms* are often useful to identify resources or express the state or result of an operation. Creatings *atoms* dynamically is not an anti-pattern by itself; however, *atoms* are not collected by the BEAM's garbage collector, so values of this type live in memory while software is executing, during its entire lifetime. Also, BEAM limits the number of *atoms* that can exist in an application (*1_048_576* by default) and each *atom* has a maximum size limited to 255 Unicode code points. For these reasons, creating *atoms* dynamically can be considered an anti-pattern in some situations, since in this way the developer has no control over how many *atoms* will be created during the software execution. This unpredictable scenario can expose the software to unexpected behavior caused by excessive memory usage, or even by reaching the maximum number of *atoms* possible.

Reviewer 5: * https://elixir-lang.org/getting-started/basic-types.html refers to `Elixir basic types`
* https://elixir-lang.org/getting-started/typespecs-and-behaviours.html refers to English-version `behaviour` (with the "u")

```

By explicitly listing all supported statuses, you guarantee only a limited number of conversions may happen. Passing an invalid status will lead to a function clause error.

An alternative is to use `String.to_existing_atom/1`, which will only convert a string to atom if the atom already exists in the system:

```elixir
defmodule MyRequestHandler do
  def parse(%{"status" => status, "message" => message} = _payload) do
    %{status: String.to_existing_atom(status), message: message}
  end
end
```

```elixir
iex> MyRequestHandler.parse(%{"status" => "status_not_seen_anywhere", "message" => "all good"})
** (ArgumentError) errors were found at the given arguments:

  * 1st argument: not an already existing atom
```

In such cases, passing an unknown status will raise as long as the status was not defined anywhere as an atom in the system. However, assuming `status` can be either `:ok`, `:error`, or `:redirect`, how can you guarantee those atoms exist? You must ensure those atoms exist somewhere **in the same module** where `String.to_existing_atom/1` is called. For example, if you had this code:

> 

Reviewer 2: ```suggestion
```
